# Task to 4.7: Deriving New Variables
# Contents
## 1. Import libraries&date sets
## 2. Step 1a. Create “price_label” column.
## 3. Step 1b. Create "busiest_day" column.
## 4. Create “Busiest days” columns
## 5. Step 3. Check values of new "busiest days" column for accuracy with observations in markdown format.
## 6. Step 4. Create a new column “busiest_period_of_day”, containings the busiest hours of the day/periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.” 
## 7 .Step 5. Print the frequency for new column "busiest_period_of_day".
## 8 .Step 7. Export your dataframe as a pickle file 

## 1. Import libraries&date sets

In [1]:
#Import libraries
import pandas as pd
import numpy as np
import os

In [3]:
# Create project folder path into a string
path=r'/Users/yevgeniyaem/Documents/Weiterbildung Data Analytics/11-2024 Instacart Basket Analysis'

In [5]:
path

'/Users/yevgeniyaem/Documents/Weiterbildung Data Analytics/11-2024 Instacart Basket Analysis'

In [7]:
# Import "orders_products_merged.pkl"
ords_prods_merged=pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

In [8]:
ords_prods_merged.describe()

user_id  customers_order_count  orders_day_of_week  \
count  3.240486e+07           3.240486e+07        3.240486e+07   
mean   1.029372e+05           1.714230e+01        2.738867e+00   
std    5.946610e+04           1.753532e+01        2.090077e+00   
min    1.000000e+00           1.000000e+00        0.000000e+00   
25%    5.142200e+04           5.000000e+00        1.000000e+00   
50%    1.026160e+05           1.100000e+01        3.000000e+00   
75%    1.543890e+05           2.400000e+01        5.000000e+00   
max    2.062090e+05           9.900000e+01        6.000000e+00   

       order_hour_of_day  days_since_prior_order    product_id  \
count       3.240486e+07            3.240486e+07  3.240486e+07   
mean        1.342515e+01            1.039267e+01  2.559866e+04   
std         4.246380e+00            8.917825e+00  1.408400e+04   
min         0.000000e+00            0.000000e+00  1.000000e+00   
25%         1.000000e+01            4.000000e+00  1.354400e+04   
50%         1.300000e+01            7.000000e+00  2.530200e+04   
75%         1.600000e+01            1.400000e+01  3.794700e+04   
max         2.300000e+01            3.000000e+01  4.968800e+04   

       add_to_cart_order     reordered      aisle_id  department_id  \
count       3.240486e+07  3.240486e+07  3.240486e+07   3.240486e+07   
mean        8.352547e+00  5.895873e-01  7.119612e+01   9.919792e+00   
std         7.127071e+00  4.919087e-01  3.821139e+01   6.281485e+00   
min         1.000000e+00  0.000000e+00  1.000000e+00   1.000000e+00   
25%         3.000000e+00  0.000000e+00  3.100000e+01   4.000000e+00   
50%         6.000000e+00  1.000000e+00  8.300000e+01   9.000000e+00   
75%         1.100000e+01  1.000000e+00  1.070000e+02   1.600000e+01   
max         1.450000e+02  1.000000e+00  1.340000e+02   2.100000e+01   

             prices  
count  3.240486e+07  
mean   1.198023e+01  
std    4.956554e+02  
min    1.000000e+00  
25%    4.200000e+00  
50%    7.400000e+00  
75%    1.130000e+01  
max    9.999900e+04

In [11]:
# Create a subset with 1mil rows
df = ords_prods_merged[:1000000]

In [ ]:
df.shape

In [13]:
df.describe()

user_id  customers_order_count  orders_day_of_week  \
count  1000000.000000         1000000.000000      1000000.000000   
mean      3295.070459              17.313612            2.765702   
std       1866.882221              17.931306            2.091348   
min          1.000000               1.000000            0.000000   
25%       1699.000000               5.000000            1.000000   
50%       3350.000000              11.000000            3.000000   
75%       4937.000000              24.000000            5.000000   
max       6474.000000              99.000000            6.000000   

       order_hour_of_day  days_since_prior_order      product_id  \
count     1000000.000000          1000000.000000  1000000.000000   
mean           13.416690               10.358285    25595.891283   
std             4.221497                8.944049    14077.246950   
min             0.000000                0.000000        1.000000   
25%            10.000000                4.000000    13536.000000   
50%            13.000000                7.000000    25340.000000   
75%            16.000000               14.000000    37825.000000   
max            23.000000               30.000000    49687.000000   

       add_to_cart_order       reordered        aisle_id   department_id  \
count     1000000.000000  1000000.000000  1000000.000000  1000000.000000   
mean            8.312942        0.590985       71.058417        9.984846   
std             7.208310        0.491652       38.228388        6.294836   
min             1.000000        0.000000        1.000000        1.000000   
25%             3.000000        0.000000       31.000000        4.000000   
50%             6.000000        1.000000       83.000000        9.000000   
75%            11.000000        1.000000      107.000000       16.000000   
max            95.000000        1.000000      134.000000       21.000000   

               prices  
count  1000000.000000  
mean        11.396794  
std        425.994349  
min          1.000000  
25%          4.200000  
50%          7.400000  
75%         11.300000  
max      99999.000000

In [17]:
df.dtypes

order_id                    object
user_id                      int64
customers_order_count        int64
orders_day_of_week           int64
order_hour_of_day            int64
days_since_prior_order     float64
product_id                   int64
add_to_cart_order            int64
reordered                    int64
product_name                object
aisle_id                     int64
department_id                int64
prices                     float64
_merge                    category
dtype: object

In [27]:
count_99999 = (df['prices'] == 99999.0).sum() 
print(f"The value 99999.0 appears {count_99999} times in the 'prices' column.")

The value 99999.0 appears 15 times in the 'prices' column.


## 2. Step 1a. Create “price_label” column.

In [15]:
# Create new "price_label" column, define a function:
def price_label(row):
    
    if row['prices'] <=5:
        return 'Low range product'
    elif (row['prices'] >5) and (row['prices'] <=15):
        return 'Medium range product'
    elif row['prices'] >15:
        return 'High range product'
    else: return np.nan

In [19]:
# Apply the function
df = df.copy()
df['price_range'] = df.apply(price_label, axis=1)

In [20]:
# Check the values in the new column 'price_range'
df['price_range'].value_counts(dropna=False)

price_range
Medium range product    673183
Low range product       314392
High range product       12425
Name: count, dtype: int64

In [23]:
# Check what the most expensive product within the subset is?
df['prices'].max()

99999.0

In [ ]:
result

In [29]:
df.columns

Index(['order_id', 'user_id', 'customers_order_count', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order', 'product_id',
       'add_to_cart_order', 'reordered', 'product_name', 'aisle_id',
       'department_id', 'prices', '_merge', 'price_range'],
      dtype='object')

In [31]:
# Create a new column "price_range_loc" with loc-function with the whole df ords_prods_merged, apply conditions for "High-tange product"
ords_prods_merged.loc[ords_prods_merged['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [32]:
# Create a new column "price_range_loc" with loc-function with the whole df ords_prods_merged, apply conditions for "Mid-range product"
ords_prods_merged.loc[(ords_prods_merged['prices'] <= 15) & (ords_prods_merged['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [35]:
# Create a new column "price_range_loc" with loc-function with the whole df ords_prods_merged, apply conditions for "Low-range product"
ords_prods_merged.loc[ords_prods_merged['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [37]:
# Find out on which day most orders take place via printing the frequency of the “orders_day_of_week” column
ords_prods_merged['orders_day_of_week'].value_counts(dropna=False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

## 3. Step 1b. Create "busiest_day" column.

In [39]:
# Create a new column “busiest day,” that will contain one of three different values: “Busiest day,” “Least busy,” and “Regularly busy".
result = []

for value in ords_prods_merged["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [40]:
# Create a new column within ords_prods_merge dataframe and set it equal to result
ords_prods_merged['Busiest_day']=result

In [43]:
ords_prods_merged['Busiest_day'].value_counts(dropna=False)

Busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

In [45]:
# Summary column for the busiest day of the week: printing the frequency of new column and cross-check it with the frequency of “orders_day_of_week"
ords_prods_merged['orders_day_of_week'].value_counts(dropna=False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

## 4. Create “Busiest days” columns = two busiest days of the week, and two slowest days colum.

In [47]:
# Create column "Busiest days" from "orders_day_of_week", to identify two busiest days (0, 1), slowest days (3, 4), or regularly busy (all other days).
result_2 = []

for value in ords_prods_merged["orders_day_of_week"]:
  if value == 0 or value == 1:
    result_2.append("Busiest day")
  elif value == 3 or value == 4:
    result_2.append("Least busy")
  else:
    result_2.append("Regularly busy")

In [48]:
result_2

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Reg

## 5. Step 3. Check values of new "busiest days" column for accuracy with observations in markdown format.

In [51]:
# Create a new column within ords_prods_merge dataframe to view results
ords_prods_merged['Busiest_days']=result_2

In [53]:
ords_prods_merged['Busiest_days'].value_counts(dropna=False)

Busiest_days
Regularly busy    12916111
Busiest day       11864412
Least busy         7624336
Name: count, dtype: int64

In [55]:
# Check output of ords_prods_merged with new "busiest days" column.
ords_prods_merged.head()

order_id  user_id  customers_order_count  orders_day_of_week  \
0  2539329        1                      1                   2   
1  2539329        1                      1                   2   
2  2539329        1                      1                   2   
3  2539329        1                      1                   2   
4  2539329        1                      1                   2   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     0.0         196                  1   
1                  8                     0.0       14084                  2   
2                  8                     0.0       12427                  3   
3                  8                     0.0       26088                  4   
4                  8                     0.0       26405                  5   

   reordered                             product_name  aisle_id  \
0          0                                     Soda        77   
1          0  Organic Unsweetened Vanilla Almond Milk        91   
2          0                      Original Beef Jerky        23   
3          0               Aged White Cheddar Popcorn        23   
4          0         XL Pick-A-Size Paper Towel Rolls        54   

   department_id  prices _merge    price_range_loc     Busiest_day  \
0              7     9.0   both  Mid-range product  Regularly busy   
1             16    12.5   both  Mid-range product  Regularly busy   
2             19     4.4   both  Low-range product  Regularly busy   
3             19     4.7   both  Low-range product  Regularly busy   
4             17     1.0   both  Low-range product  Regularly busy   

     Busiest_days  
0  Regularly busy  
1  Regularly busy  
2  Regularly busy  
3  Regularly busy  
4  Regularly busy

In [57]:
# Check df ords_pords_merged
ords_prods_merged.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 17 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   order_id                object  
 1   user_id                 int64   
 2   customers_order_count   int64   
 3   orders_day_of_week      int64   
 4   order_hour_of_day       int64   
 5   days_since_prior_order  float64 
 6   product_id              int64   
 7   add_to_cart_order       int64   
 8   reordered               int64   
 9   product_name            object  
 10  aisle_id                int64   
 11  department_id           int64   
 12  prices                  float64 
 13  _merge                  category
 14  price_range_loc         object  
 15  Busiest_day             object  
 16  Busiest_days            object  
dtypes: category(1), float64(2), int64(9), object(5)
memory usage: 3.9+ GB


### The total value counts for the ords_prods_merged dataframe is equal to the total sum after update of all busy days categories ("Regularly busy", "Busiest days", "Slowest days"). Also, the updated value counts match the sum totals for defined days of the week (e.g. "Busiest days" = 11864412, which matches orders_day_of_week 0 + 1, or 6204182 + 5660230).

## 6. Step 4. Create a new column “busiest_period_of_day”, containings the busiest hours of the day/periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.” 

In [59]:
# Find out on which hour most orders take place via printing the frequency of the “order_hour_of_day” column
ords_prods_merged['order_hour_of_day'].value_counts(dropna=False)

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

In [73]:
# Create column "busiest_period_of_day" from "order_hour_of_day", to identify periods of 
# "Most orders" = 10, 11, 14, 15, 13, 12, 16, 9; 
# "Average orders" = 17, 8, 18, 19, 20, 7, 21, 22;
# and "Fewest orders" = 23, 6, 0, 1, 5, 2, 4, 3.
result_3 = []

for value in ords_prods_merged["order_hour_of_day"]:
  if value in [10, 11, 14, 15, 13, 12, 16, 9]:
    result_3.append("Most orders")
  elif value in [23, 6, 0, 1, 5, 2, 4, 3]:
    result_3.append("Fewest orders")
  else:
    result_3.append("Average orders")

In [74]:
result_3

['Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Avera

In [79]:
# Create new column "busiest_period_of_day" in ords_prods_merge.
ords_prods_merged['busiest_period_of_day'] = result_3

In [81]:
# Check output of ords_prods_merged with new "busiest_period_of_day" column.
ords_prods_merged.head()

order_id  user_id  customers_order_count  orders_day_of_week  \
0  2539329        1                      1                   2   
1  2539329        1                      1                   2   
2  2539329        1                      1                   2   
3  2539329        1                      1                   2   
4  2539329        1                      1                   2   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     0.0         196                  1   
1                  8                     0.0       14084                  2   
2                  8                     0.0       12427                  3   
3                  8                     0.0       26088                  4   
4                  8                     0.0       26405                  5   

   reordered                             product_name  aisle_id  \
0          0                                     Soda        77   
1          0  Organic Unsweetened Vanilla Almond Milk        91   
2          0                      Original Beef Jerky        23   
3          0               Aged White Cheddar Popcorn        23   
4          0         XL Pick-A-Size Paper Towel Rolls        54   

   department_id  prices _merge    price_range_loc     Busiest_day  \
0              7     9.0   both  Mid-range product  Regularly busy   
1             16    12.5   both  Mid-range product  Regularly busy   
2             19     4.4   both  Low-range product  Regularly busy   
3             19     4.7   both  Low-range product  Regularly busy   
4             17     1.0   both  Low-range product  Regularly busy   

     Busiest_days busiest_period_of_day  
0  Regularly busy        Average orders  
1  Regularly busy        Average orders  
2  Regularly busy        Average orders  
3  Regularly busy        Average orders  
4  Regularly busy        Average orders

## 7 .Step 5. Print the frequency for new column "busiest_period_of_day".

In [83]:
# Print value counts in "busiest_period_of_day" column.
ords_prods_merged['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most orders       21118071
Average orders     9997651
Fewest orders      1289137
Name: count, dtype: int64

## 8 .Step 7. Export your dataframe as a pickle file (since you added new columns) and store it correctly in your “Prepared Data” folder.

In [85]:
# Export of df ords_prods_merged in pickle format as “orders_products_merged_derived.pkl”.
ords_prods_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_derived.pkl'))